md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2326 @@
+format, positions are represented with a special syntax.
+
+#### Data Section
+
+**Name:** `data`
+
+The Data Section consists of an [array] of data initializers.
+
+A *data initializer* consists of a [string] and a *start offset*. It describes
+data to be loaded into linear memory as part of
+[linear memory instantiation](#linear-memory-instantiation).
+
+**Validation:**
+ - For each element of the array, the sum of the start offset and the length of
+   the string is required to be less than the initial size declared in the
+   [Memory Section](#memory-section).
+ - The start offset of each element of the array is required to be greater
+   than the index of any byte initialized by elements that precede it.
+
+#### Name Section
+
+**Name:** `name`
+
+The Names Section doesn't change execution semantics and malformed constructs,
+such as out-of-bounds indices, in this section cause the section to be ignored,
+and don't trigger validation failures.
+
+The Names Section consists of an [array] of function name descriptors, which
+each describe names for the function with the corresponding index in the module,
+and which consist of:
+ - the function name, a [string].
+ - the names of the locals in the function, an [array] of [strings].
+
+> Name data is represented as an explicit section in the binary format, however
+in the text format it may be represented as an integrated part of the syntax for
+functions rather than as a discrete section.
+
+> The expectation is that, when a binary WebAssembly module is presented in a
+human-readable format in a browser or other development environment, the names
+in this section are to be used as the names of functions and locals in the
+[text format](TextFormat.md).
+
+TODO: Should the names in this section be required to be valid UTF-8 strings?
+This section isn't used during normal validation or execution, so it's off the
+"hot path" and is only used during debugging, to present strings to humans, so
+it might make sense.
+
+
+Validation
+--------------------------------------------------------------------------------
+
+0. [Module Validation](#module-validation)
+0. [Function Validation](#function-validation)
+
+### Module Validation
+
+Validation of a module entails performing the **Validation** clause of the
+[top-level module description](#module-contents).
+
+Then, if the module contains a [Code Section](#code-section), each function body
+in the section is [validated](#function-validation).
+
+### Function Validation
+
+The major requirements for function validation are:
+ - Control-flow constructs must form properly nested *regions*. `loop`, `block`,
+   and the function entry pair with `end`, and `if` can pair with `end` or
+   `else`, which is then paired with `end`.
+ - The number and types of all elements that can be on the value stack at
+   execution of each instruction is required be the same for all control-flow
+   paths to that instruction.
```

Comment:
Was pondering if the order needs to be the same too? Does this clause make a `swap` operator problematic?

Also it seems that the one exception would be setting a stack element to a value that could not be consumed because if it were then it would be statically detected and the code would not validate? This seems to be permitting the `drop` operator.